EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/AccountStorage.java

```diff
@@ -0,0 +1,133 @@
+package net.essentialsx.discordlink;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.Writer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class AccountStorage {
+    private final static Logger logger = Logger.getLogger("EssentialsDiscordLink");
+    private final Gson gson = new Gson();
+    private final EssentialsDiscordLink ess;
+    private final File accountFile;
+    private final ConcurrentHashMap<String, String> uuidToDiscordIdMap;
+    private final AtomicBoolean mapDirty = new AtomicBoolean(false);
+    private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+
+    public AccountStorage(final EssentialsDiscordLink ess) throws IOException {
+        this.ess = ess;
+        this.accountFile = new File(ess.getDataFolder(), "accounts.json");
+        if (!ess.getDataFolder().exists() && !ess.getDataFolder().mkdirs()) {
+            throw new IOException("Unable to create account file!");
+        }
+        if (!accountFile.exists() && !accountFile.createNewFile()) {
+            throw new IOException("Unable to create account file!");
+        }
+        try (final Reader reader = new FileReader(accountFile)) {
+            //noinspection UnstableApiUsage
+            final Map<String, String> map = gson.fromJson(reader, new TypeToken<Map<String, String>>() {}.getType());
+            uuidToDiscordIdMap = map == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(map);
+        }
+
+        executorService.scheduleAtFixedRate(() -> {
+            if (!mapDirty.compareAndSet(true, false)) {
+                return;
+            }
+
+            if (ess.getEss().getSettings().isDebug()) {
+                logger.log(Level.INFO, "Saving linked discord accounts to disk...");
+            }
+
+            final Map<String, String> clone;
+            synchronized (uuidToDiscordIdMap) {
+                clone = new HashMap<>(uuidToDiscordIdMap);
+            }
+            try (final Writer writer = new FileWriter(accountFile)) {
+                gson.toJson(clone, writer);
+            } catch (IOException e) {
+                logger.log(Level.SEVERE, "Failed to save link accounts!", e);
+                mapDirty.set(true); // mark the map as dirty and pray it fixes itself :D
+            }
+        }, 0, 10, TimeUnit.SECONDS);
+    }
+
+    public void add(final UUID uuid, final String discordId) {
```

Comment:
How are MC accounts mapped to Discord? If it's supposed to be one-to-one, it may be worth putting constraints on this (or even just using a `BiMap`). If it's *not* one-to-one, we should probably give users all the current accounts they're linked to. 

I'd prefer one-to-one as it makes the account link system far easier to understand, but this would prevent people from linking alts without matching Discord alts.

Reply:
no it's not